### PR TITLE
Replay pivot edits

### PR DESCRIPTION
**Pull request by naterush**
# Description

Now, when you make a pivot, we replay edits on top of it. 

Note, code optimizing with all this isn't perfect, and it's possible to get not perfectly optimized code. But I think for realistic use cases, this is going to work real well!

# Testing

Make a pivot table, make some edits on top of it (like adding columns), and then edit the pivot table. Then make some more edits (like filtering, idk), and then edit the pivot table again. 

Mito will keep all of your edits that don't error, which is pretty sweet. 

See both backend tests and frontend tests for pivot. 

# Documentation

Note the internal documentation. This approach is a bit of a hack, and we pick up some complexity -- but overall it seems pretty robust.

### CODE REVIEW

File: mitosheet/mitosheet/code_chunks/code_chunk.py [hunk 1]
```diff
@@ -49,7 +49,15 @@ def get_code(self) -> Tuple[List[str], List[str]]:
...
         """
         raise NotImplementedError('Implement in subclass')
-
+    
+    def get_optional_code_lines(self) -> List[str]:
+        """
+        Returns a list of extra code that is applied after this code chunk, for example
+        the code that runs in the pivot table. Notably, these are _optional_ code lines,
+        as they are not required to run the code chunk.
+        """
```
Comment (naterush) [inline mitosheet/mitosheet/code_chunks/code_chunk.py:58]: Add documentation here.

File: mitosheet/mitosheet/code_chunks/step_performers/pivot_code_chunk.py [hunk 2]
```diff
@@ -104,7 +104,8 @@ def __init__(
         values_column_ids_map: Dict[ColumnID, Collection[str]],
         flatten_column_headers: Optional[bool],
         public_interface_version: int,
-        new_df_name: str
+        new_df_name: str,
+        executed_optional_lines: Optional[List[str]] = None
```
Comment (naterush) [inline mitosheet/mitosheet/code_chunks/step_performers/pivot_code_chunk.py:108]: Docs for this

File: mitosheet/mitosheet/step.py [hunk 3]
```diff
@@ -149,13 +150,56 @@ def set_prev_state_and_execute(self, new_prev_state: State) -> bool:
...
+
         # Actually execute the data transformation
         post_state_and_execution_data = self.step_performer.execute(new_prev_state, params)
 
         if post_state_and_execution_data is not None:
             # If we don't get anything new back, then we just make this
             # step a no-op, and don't do anything
             (new_post_state, execution_data) = post_state_and_execution_data
+
+            # Another option is that we exec this code here. The benefit of this
```
Comment (naterush) [inline mitosheet/mitosheet/step.py:197]: Remove this

File: mitosheet/mitosheet/step.py [hunk 4]
```diff
@@ -149,13 +150,56 @@ def set_prev_state_and_execute(self, new_prev_state: State) -> bool:
         # we can get all of it without executing. I think we probably can
         params = self.step_performer.saturate(new_prev_state, self.params)
 
+        # If this is a pivot step, that has a destination sheet index, then we're going to 
+        # go look for the original step that created this pivot
+        if self.step_type == 'pivot' and params.get('destination_sheet_index') is not None:
```
Comment (naterush) [inline mitosheet/mitosheet/step.py:155]: Factor this into a utility

File: mitosheet/mitosheet/step_performers/pivot.py [hunk 5]
```diff
@@ -100,6 +100,8 @@ def saturate(cls, prev_state: State, params: Dict[str, Any]) -> Dict[str, Any]:
     def execute(cls, prev_state: State, params: Dict[str, Any]) -> Tuple[State, Optional[Dict[str, Any]]]:
         sheet_index: int = get_param(params, 'sheet_index')
         destination_sheet_index: Optional[int] = get_param(params, 'destination_sheet_index')
+
+        print("CODE LINES", params.get('extra_code'))
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/pivot.py:104]: Remove

File: mitosheet/mitosheet/step_performers/pivot.py [hunk 6]
```diff
@@ -110,6 +112,7 @@ def execute(cls, prev_state: State, params: Dict[str, Any]) -> Tuple[State, Opti
             # actually used by the backend (which is kinda confusing and should be renamed)
             'destination_sheet_index': final_destination_sheet_index,
             'new_df_name': new_df_name,
+            'extra_code': params.get('extra_code')
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/pivot.py:115]: Should not put this in the execution data - I think this should just be a straight up param. We duplicate some work here

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 7]
```diff
@@ -149,9 +150,31 @@ def execute_through_transpile(
         # but in practice is seems to work...
         exec_globals = get_globals_for_exec(post_state, post_state.public_interface_version)
         exec_locals = {**exec_globals}
+
+        print("exec_globals", exec_globals.get('df1_pivot'))
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/step_performer.py:154]: Remove prints here

File: mitosheet/mitosheet/steps_manager.py [hunk 8]
```diff
@@ -96,7 +96,7 @@ def execute_step_list_from_index(
 
         # Set the previous state of the new step, and then update
         # what the last valid step is
-        new_step.set_prev_state_and_execute(last_valid_step.final_defined_state)
+        new_step.set_prev_state_and_execute(last_valid_step.final_defined_state, new_step_list)
```
Comment (naterush) [inline mitosheet/mitosheet/steps_manager.py:99]: This feels like the wrong variable to pass

**Review by naterush — COMMENTED** [review]
Let's get some documentation!

**Comment by naterush** [comment]
Handle the bug where we add extra comments and whitespace at the bottom

File: mitosheet/mitosheet/api/get_test_imports.py [hunk 9]
```diff
@@ -43,7 +43,9 @@ def get_test_imports(params: Dict[str, Any], steps_manager: StepsManagerType) ->
     for index, _import in enumerate(all_imports):
         try:
             step = Step(_import['step_type'], 'fake_id', _import["params"])
-            executed = step.set_prev_state_and_execute(steps_manager.steps_including_skipped[0].final_defined_state)
+            # We don't need the previous steps for this, so we just pass an empty list
+            # to this parameter. We only need this parameter for pivot tables
```
Comment (naterush) [inline mitosheet/mitosheet/api/get_test_imports.py:47]: Update comment

File: mitosheet/mitosheet/code_chunks/code_chunk.py [hunk 10]
```diff
@@ -49,7 +51,41 @@ def get_code(self) -> Tuple[List[str], List[str]]:
...
+
+        Thus, we have a new type of code for the code chunk. The `get_code` results _must_
+        all execute correctly, for the step to terminate correctly. If any of the lines of
+        code from get_code fail, then the entire step is rolled back and not created. 
+
+        However, optional code lines can fail and the step still be correctly executed. Thus, 
+        we have to have a notion of a) optional code lines, and b) the code chunk needs to know
+        which of these optional lines was executed correctly. 
+
+        In the above example, fi the rename step failed, then this function would return the
```
Comment (naterush) [inline mitosheet/mitosheet/code_chunks/code_chunk.py:84]: Typo

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 11]
```diff
@@ -150,6 +154,27 @@ def execute_through_transpile(
...
+
+        optional_code_that_successfully_executed = []
+        for optional_code_line in optional_code or []:
+
+            # We don't need to exec spaces
+            if optional_code_line == '':
+                optional_code_that_successfully_executed.append(optional_code_line)
+                continue
+
+            # TODO: we should make it so it rolls back the state if this fails
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/step_performer.py:166]: Update

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 12]
```diff
@@ -150,6 +154,27 @@ def execute_through_transpile(
...
+            # because we 
+            try:
+                exec(optional_code_line, exec_globals, exec_locals)
+                optional_code_that_successfully_executed.append(optional_code_line)
+            except Exception as e:
+
+                # TODO: If we hit an error, we remove lines up to last non-comment
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/step_performer.py:173]: TODO

**Review by naterush — COMMENTED** [review]
Cleanup

File: mitosheet/mitosheet/tests/step_performers/test_melt.py [hunk 13]
```diff
@@ -69,6 +69,8 @@ def test_melt(input_dfs, sheet_index, id_vars, value_vars, output_dfs):
 
     assert len(mito.dfs) == len(output_dfs)
     for actual, expected in zip(mito.dfs, output_dfs):
+        print(actual)
+        print(expected)
```
Comment (naterush) [inline mitosheet/mitosheet/tests/step_performers/test_melt.py:73]: Remove

**Review by naterush — COMMENTED** [review]
Cleanup

**Comment by aarondr77** [comment]
1. Import the loans dataset
2. Create a pivot table with `issue date` as the row and add a value
3. Change the dtype of `issue date` to `datetime`
4. Add a new column to the values section 
5. See that the generated code has `import pandas as pd` as the last line

**Comment by aarondr77** [comment]
1. Import 
2. Create pivot table
3. Filter on one of the values columns
4. Add a new column to the values section
5. Notice that the data is still filtered (in the spreadsheet and the generated code), but the filter icon is not turned on and if you open the control panel, there is not a filter applied.

**Comment by aarondr77** [comment]
We don't need to address this now, but let's keep an eye out https://github.com/mito-ds/mito/issues/1107

**Review by aarondr77 — COMMENTED** [review]
This code is nice. I like that the same approach is extendable to other use cases and you wrote a bunch of tests. 

(In the spirit of moving away from success theatre), it too soon to tell if this feature is going to be useful. My hunch is that it will be and users won't get tripped up by specific edge cases not being replayed. They won't expect it to be. 

I'm excited to watch some users on real workflows use this and evaluate how well it works.

**Comment by aarondr77** [comment]
Can we add a `edit pivot table` button to the pivot table sheet tab. I found myself wanting that often while testing

File: mitosheet/mitosheet/step_performers/pivot.py [hunk 14]
```diff
@@ -77,6 +90,13 @@ def saturate(cls, prev_state: State, params: Dict[str, Any]) -> Dict[str, Any]:
                     new_aggregation_function_names.append(i)
             params['values_column_ids_map'][column_id] = new_aggregation_function_names
 
+        # Add in any optional code that we want to try and execute
+        # which in this case are the edits we replay on top of the 
+        # pivot table
+        optional_code = get_optional_code_to_replay_on_pivot(previous_steps, params)
+        print("SATURATING", previous_steps, optional_code)
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/pivot.py:97]: Remove

**Review by naterush — COMMENTED** [review]
Minor cleanup, then LGTM

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 15]
```diff
@@ -150,6 +154,29 @@ def execute_through_transpile(
...
+        non_code_lines_before_optional_line = []
+        for optional_code_line in optional_code or []:
+
+            # We don't need to exec spaces
+            if optional_code_line == '' or optional_code_line.strip().startswith('#'):
+                non_code_lines_before_optional_line.append(optional_code_line)
+                continue
+
+            # TODO: we should make it so it rolls back the state if this fails
+            # but it's fine for now -- since partial updates don't seem to
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/step_performers/step_performer.py]: Just manual testing. I think this is fine to leave for now.

**Comment by aarondr77** [comment]
Can we add a edit pivot table button to the pivot table sheet tab. I found myself wanting that often while testing

**Comment by aarondr77** [comment]
> Note, code optimizing with all this isn't perfect, and it's possible to get not perfectly optimized code. But I think for realistic use cases, this is going to work real well!

Currently it doesn't look like any code optimization is happening. Here's a few tests I did:

Add + Rename
1. Create a pivot table
2. Add a column to the pivot table + rename it
3. Edit the pivot table configuration
4. Notice that the add + rename code is not optimized. 

Add + Rename + Rename 
1. Create a pivot table
2. Add a column to the pivot table + rename it + rename it again
3. Edit the pivot table configuration
4. Notice that there are 2 consecutive renames 

Add + Delete column
1. Create a pivot table
2. Add a column
3. Delete it
4. Edit the pivot table configuration 
5. Notice that there is add and a delete code

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 16]
```diff
@@ -102,6 +105,7 @@ def execute_through_transpile(
         execution_data: Optional[Dict[str, Any]]=None,
         new_dataframe_params: Optional[ExecuteThroughTranspileNewDataframeParams]=None,
         column_headers_to_column_ids: Optional[Dict[ColumnHeader, ColumnID]]=None,
+        optional_code: Optional[Tuple[List[str], List[str]]]=None,
```
Comment (aarondr77) [inline mitosheet/mitosheet/step_performers/step_performer.py:108]: See https://github.com/mito-ds/mito/issues/1110

**Review by aarondr77 — COMMENTED** [review]
todo

File: mitosheet/mitosheet/step_performers/pivot.py [hunk 17]
```diff
@@ -155,4 +179,60 @@ def get_new_pivot_df_name(prev_state: State, sheet_index: int) -> str:
...
+    optional_code: Tuple[List[str], List[str]] = ([], [])
+
+    # First, find the index of the step we have to start looking from
+    # for steps that edit the pivot table
+    starting_index = None
+    for index, step in enumerate(reversed(previous_steps)):
+        # Case 1: we find the last step that edited this pivot, and get it's extra code 
+        # that successfully executed, as we want to start by replaying this
+        if step.step_type == 'pivot' and 'destination_sheet_index' in step.params and step.params['destination_sheet_index'] == destination_sheet_index:
+            optional_code = step.execution_data.get('optional_code_that_successfully_executed', ([], []))
```
Comment (aarondr77) [inline mitosheet/mitosheet/step_performers/pivot.py:206]: There's no reason to do this in the first section of this function. Let's move it down to the next section

`previous_steps = previous_steps[starting_index + 1:]` -> `previous_steps = previous_steps[starting_index:]`

File: mitosheet/mitosheet/code_chunks/code_chunk.py [hunk 18]
```diff
@@ -49,7 +51,41 @@ def get_code(self) -> Tuple[List[str], List[str]]:
...
+
+        Thus, we have a new type of code for the code chunk. The `get_code` results _must_
+        all execute correctly, for the step to terminate correctly. If any of the lines of
+        code from get_code fail, then the entire step is rolled back and not created. 
+
+        However, optional code lines can fail and the step still be correctly executed. Thus, 
+        we have to have a notion of a) optional code lines, and b) the code chunk needs to know
+        which of these optional lines was executed correctly. 
+
+        In the above example, if the rename step failed, then this function would return the
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/code_chunks/code_chunk.py:84]: Everything on top of the pivot table is optional! Optional lines just means that as soon as one optional line fails, the rest of them stop executing.

File: mitosheet/mitosheet/step_performers/step_performer.py (hunk 16 above)
Comment (naterush) [reply mitosheet/mitosheet/step_performers/step_performer.py:108]: This is fixed in #1115

File: mitosheet/mitosheet/step_performers/pivot.py (hunk 17 above)
Comment (naterush) [reply mitosheet/mitosheet/step_performers/pivot.py:206]: I don't understand this comment.

**Comment by naterush** [comment]
@aarondr77 not adding a pivot edit button, because this functionality is already extremely accessible through a few other locations:
1. Open the pivot table
2. The sheet tab context menu
3. Clicking the pivot button with a pivot table open
